src/reflect/scala/reflect/internal/transform/Erasure.scala

```diff
@@ -112,8 +112,9 @@ trait Erasure {
     protected def eraseDerivedValueClassRef(tref: TypeRef): Type = erasedValueClassArg(tref)
 
     def apply(tp: Type): Type = tp match {
-      case ConstantType(_) =>
-        tp
+      case ConstantType(ct) =>
+        if (ct.tag == ClazzTag) ConstantType(Constant(apply(ct.typeValue)))
+        else tp
```

Comment:
OK I can reproduce it. It happens precisely for `final val`s whose right-hand-side is a `classOf[Unit]`. Compiling
```scala
object Foo {
  final val C = classOf[Unit]
}
```
gives, after erasure:
```scala
  object Foo extends Object {
    def <init>(): helloworld.Foo.type = {
      Foo.super.<init>();
      ()
    };
    final private[this] val C: Class(classOf[scala.runtime.BoxedUnit]) = classOf[scala.runtime.BoxedUnit];
    final <stable> <accessor> def C(): Class(classOf[scala.runtime.BoxedUnit]) = classOf[scala.runtime.BoxedUnit]
  };
```